README.md
# [is](https://github.com/jsmini/is)

[![](https://img.shields.io/badge/Powered%20by-jslib%20base-brightgreen.svg)](https://github.com/yanhaijing/jslib-base)
[![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/jsmini/is/blob/master/LICENSE)
[![CI](https://github.com/jsmini/is/actions/workflows/ci.yml/badge.svg?branch=master)](https://github.com/jsmini/is/actions/workflows/ci.yml)
[![npm](https://img.shields.io/badge/npm-0.10.0-orange.svg)](https://www.npmjs.com/package/@jsmini/is)
[![NPM downloads](http://img.shields.io/npm/dm/@jsmini/is.svg?style=flat-square)](http://www.npmtrends.com/@jsmini/is)
[![Percentage of issues still open](http://isitmaintained.com/badge/open/jsmini/is.svg)](http://isitmaintained.com/project/jsmini/is 'Percentage of issues still open')

js判断函数，解决80%类型判断问题，原生兼容IE6

Type checking function. fixed 80% of type checking problem

English | [简体中文](./README-zh_CN.md)

## Environment Support

unit test ensure it supports the following environments.

| IE/Edge | Chrome | Firefox | Safari | Opera | IOS | Android | Node  |
| ------- | ------ | ------- | ------ | ----- | --- | ------- | ----- |
| 6+      | 23+    | 4+      | 6+     | 10+   | 5+  | 2.3+    | 0.10+ |

## Directory

```
.
├── demo
├── dist  # production code
├── doc   # document
├── src   # source code
├── test  # unit test
├── CHANGELOG.md
└── TODO.md
```

## Usage

npm installation

```bash
$ npm install --save @jsmini/is
```

Node.js

```js
var isNumber = require('@jsmini/is').isNumber;

isNumber(1); // true
```

webpack

```js
import { isNumber } from '@jsmini/is';

isNumber(1); // true
```

Require.js

```js
requirejs(['node_modules/@jsmini/is/dist/index.aio.js'], function (jsmini_is) {
  var isNumber = jsmini_is.isNumber;

  isNumber(1); // true
});
```

Browser

```html
<script src="node_modules/@jsmini/is/dist/index.aio.js"></script>

<script>
  var isNumber = jsmini_is.isNumber;

  isNumber(1); // true
</script>
```

## Document

- [API](https://github.com/jsmini/is/blob/master/doc/api.md)

## Contributing Guide ![PRs welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)

when initialize, install dependencies

```bash
$ npm install
```

builds your code for production to `build` folder

```bash
$ npm run build
```

run unit test. notice: borwser enviroment need to test manually. test file is in `test/browser`

```bash
$ npm test
```

change the version in package.json and README.md, add your description in CHANGELOG.md, and then release it happily.

```bash
$ npm run release
```

publish the new package to npm

```bash
$ npm publish --access=public
```

Maybe you should change relative information in the follow files

- README.md
- package.json
- config/rollup.js
- test/browser/index.html

```bash
$ npm run rename # rename command
```

## Contributors

[contributors](https://github.com/jsmini/is/graphs/contributors)

## CHANGELOG

[CHANGELOG.md](https://github.com/jsmini/is/blob/master/CHANGELOG.md)

## TODO

[TODO.md](https://github.com/jsmini/is/blob/master/TODO.md)

## who is using


CHANGELOG.md
# 变更日志

## 0.10.0 / 2023-11-19

- 升级 jslib-base 2.3.2
- 支持 sourceMap
- 升级 @jsmini/type 0.11.0
- fix: 支持 node >= 14.0.0

## 0.9.0 / 2023-9-24

- 升级最新版 jslib-base
- 支持 Node.js ESM
- 升级 @jsmini/type

## 0.8.5 / 2019-10-10

- fix: 修复丢失d.ts的问题

## 0.8.2 / 2019-3-20

- fix: fix ts bug

## 0.8.1 / 2019-3-4

- fix: 修复依赖不自动更新的问题

## 0.8.0 / 2019-3-2

- 增加.d.ts文件，支持ts调用
- `isEmptyString` 方法使用 `RegExp.prototype.test` 替换 `replace`

## 0.7.1 / 2018-10-10

- 修复 `isObject(Object.create(null))` 为`false`的错误

## 0.7.0 / 2018-9-29

- 新增 isInRange()
- 新增 isEmptyString()

## 0.6.0 / 2018-9-28

- isInt -> isInteger
- isInt改为代表整形

## 0.5.0 / 2018-9-27

- 新增 isInt()
- isNumber新增最小和最大值参数

## 0.4.0 / 2018-4-6

- 迁移项目，更改名称

## 0.3.0 / 2018-3-32

- 去掉babel编译

## 0.2.0 / 2018-3-31

- 原生支持ie6-8

## 0.1.0 / 2018-3-17

- 新增 isNumber()
- 新增 isBoolean()
- 新增 isString()
- 新增 isNull()
- 新增 isUndefined()
- 新增 isObject()
- 新增 isFunction()
- 新增 isArray()


test/test.js
var expect = require('expect.js');

var is = require('../src/index.js');

describe('单元测试', function() {
    this.timeout(1000);

    describe('功能', function() {
        it('相等', function() {
            expect(is.isBoolean(true)).to.equal(true);
            expect(is.isString('')).to.equal(true);
            expect(is.isNull(null)).to.equal(true);
            expect(is.isUndefined()).to.equal(true);

            expect(is.isObject({})).to.equal(true);
            expect(is.isObject(Object.create(null))).to.equal(true);
            expect(is.isObject(Object.create({}))).to.equal(true);
            
            expect(is.isArray([])).to.equal(true);
            expect(is.isFunction(function () {})).to.equal(true);
        });

        it('isInRange', function() {
            expect(is.isInRange(1)).to.equal(true);
            expect(is.isInRange('abc')).to.equal(false);

            expect(is.isInRange(1, 1)).to.equal(true);
            expect(is.isInRange(1, 2)).to.equal(false);

            expect(is.isInRange(1, undefined, 1)).to.equal(true);
            expect(is.isInRange(1, 0, 2)).to.equal(true);
            expect(is.isInRange(1, 0, 0)).to.equal(false);
        });

        it('isNumber', function() {
            expect(is.isNumber(1)).to.equal(true);
            expect(is.isNumber(1.1)).to.equal(true);
            expect(is.isNumber(1, 0, 10)).to.equal(true);
            expect(is.isNumber(1, 1, 1)).to.equal(true);
            expect(is.isNumber(1, 2)).to.equal(false);
            expect(is.isNumber(1, -1, 0)).to.equal(false);
        });
        
        it('isInteger', function() {
            expect(is.isInteger(1)).to.equal(true);
            expect(is.isInteger(1.1)).to.equal(false);
            expect(is.isInteger(1, 0, 10)).to.equal(true);
            expect(is.isInteger(1, 1, 1)).to.equal(true);
            expect(is.isInteger(1, 2)).to.equal(false);
            expect(is.isInteger(1, -1, 0)).to.equal(false);
        });

        it('isInt', function() {
            expect(is.isInt(1)).to.equal(true);
            expect(is.isInt(-2147483648)).to.equal(true);
            expect(is.isInt(-2147483649)).to.equal(false);
            expect(is.isInt(2147483647)).to.equal(true);
            expect(is.isInt(2147483648)).to.equal(false);
        });

        it('isEmptyString', function() {
            expect(is.isEmptyString(123)).to.equal(false);
            expect(is.isEmptyString('')).to.equal(true);
            expect(is.isEmptyString(' ')).to.equal(true);
            expect(is.isEmptyString('   ')).to.equal(true);
            expect(
                is.isEmptyString(`
            `)
            ).to.equal(true);
            expect(is.isEmptyString(' a ')).to.equal(false);
        });
    });
});


config/rollup.js
var babel = require('rollup-plugin-babel');

var pkg = require('../package.json');

var version = pkg.version;

var banner = 
`/*!
 * ${pkg.name} ${version} (https://github.com/jsmini/is)
 * API https://github.com/jsmini/is/blob/master/doc/api.md
 * Copyright 2017-${(new Date).getFullYear()} jsmini. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/is/blob/master/LICENSE)
 */
`;

function getCompiler(opt) {
    return babel({
        babelrc: false,
        presets: [
            [
                '@babel/preset-env',
                {
                    'targets': {
                        'browsers': 'last 2 versions, > 1%, ie >= 6, Android >= 4, iOS >= 6, and_uc > 9',
                        'node': '0.10'
                    },
                    'modules': false,
                    'loose': false
                }
            ]
        ],
        plugins: [
            [
                '@babel/plugin-transform-runtime',
                {
                    'helpers': false,
                    'regenerator': false
                }
            ]
        ],
        runtimeHelpers: true,
        exclude: 'node_modules/**'
    });
}

exports.name = 'jsmini_is';
exports.banner = banner;
exports.getCompiler = getCompiler;


src/index.js
import { type } from '@jsmini/type';

export function isInRange(x, min, max) {
  x = +x;
  min = +min;
  max = +max;

  // x 不是数字，则返回false
  if (isNaN(x)) return false;

  // min 或 max 不传，则认为不设置下限或上限
  return (!isNaN(min) ? x >= min : true) && (!isNaN(max) ? x <= max : true);
}

export function isNumber(x, min, max) {
  return type(x) === 'number' && isInRange(x, min, max);
}

export function isInteger(x, min, max) {
  return parseInt(x, 10) === x && isInRange(x, min, max);
}

export function isInt(x) {
  // -2^31 ~ 2^31-1
  return isInteger(x, -2147483648, 2147483647);
}

export function isBoolean(x) {
  return type(x) === 'boolean';
}

export function isString(x) {
  return type(x) === 'string';
}

export function isEmptyString(x) {
  if (!isString(x)) return false;

  return /^\s*$/.test(x);
}

export function isNull(x) {
  return type(x) === 'null';
}

export function isUndefined(x) {
  return type(x) === 'undefined';
}

export function isObject(x) {
  return type(x) === 'object';
}

export function isFunction(x) {
  return type(x) === 'function';
}

export const isArray = isFunction(Array.isArray)
  ? Array.isArray
  : function isArray(x) {
      return type(x) === 'array';
    };
